// mobx-remotedev/lib/utils.js
'use strict';

exports.__esModule = true;
exports.setValue = exports.silently = undefined;
exports.createAction = createAction;
exports.getName = getName;

var _mobx = require('mobx');

var _mobx2 = _interopRequireDefault(_mobx);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getPayload = function getPayload(change) {
  var added = change.added;
  var addedCount = change.addedCount;
  var index = change.index;
  var removed = change.removed;
  var removedCount = change.removedCount;

  return {
    index: index,
    added: added && _mobx2.default.toJS(added),
    addedCount: addedCount,
    removed: removed && _mobx2.default.toJS(removed),
    removedCount: removedCount
  };
};

function createAction(name, change) {
  if (!change) {
    // is action
    return { type: name };
  }

  var action = void 0;
  if (typeof change.newValue !== 'undefined') {
    var _action;

    var key = typeof change.index !== 'undefined' ? change.index : change.name;
    action = (_action = {}, _action[key] = _mobx2.default.toJS(change.newValue), _action);
  } else {
    action = getPayload(change);
  }
  action.type = '┃ ' + name;

  return action;
}

function getName(obj) {
  if (!obj || !_mobx2.default.isObservable(obj)) return '';
  var r = _mobx2.default.extras.getDebugName(obj);
  var end = r.indexOf('.');
  if (end === -1) end = undefined;
  return r.substr(0, end);
}

/* eslint-disable no-param-reassign */
var silently = exports.silently = function silently(fn, store) {
  store.__isRemotedevAction = true;
  fn();
  delete store.__isRemotedevAction;
};

function setValueAction(store, state) {
  silently(function () {
    if (store.importState) {
      store.importState(state);
    } else {
      Object.keys(state).forEach(function (key) {
        store[key] = state[key];
      });
    }
  }, store);
  return state;
}
setValueAction.__isRemotedevAction = true;
var setValue = exports.setValue = _mobx2.default.action('@@remotedev', setValueAction);
/* eslint-enable */
